### ISSUE TITLE
Possible to fall back to in-memory storage?

### ISSUE THREAD

**Issue by jonathanq** [alisaifee/flask-limiter#16]
Is it possible currently to have it fall back to an in-memory storage scheme if the redis/memcached server was down?  I didn't see anything in the code, so I assume the answer is "no".  But just wanted to ask.

I would assume this would be opt-in, not everyone would want this behavior.  But I see a huge benefit to having the rate limiting just flip back to in-memory if the remote server was down.   

If the flask app was under high load, the request load could take down a memcache/redis server.  So having it flip to in-memory means an endpoint could stay up and then the in-memory rate limit would get exceeded on each server and the offending client would get blocked (just would take longer as they'd have to hit limits on each individual server).  While other clients would still get access.

**Comment by peplin** [comment]
I have a question similar to this issue - if Redis is not up, our app hangs on startup. Is there a way to at least set a timeout?

**Comment by alisaifee** [comment]
Hi @peplin. I just tested using a configuration that points to a port that isn't serving redis and I get a `ConnectionError` when the `Limiter` extension is initialized with the app. Could you elaborate more on this hanging behavior?

**Comment by alisaifee** [comment]
@jonathanq: I've done some work on this proposal (still WIP) and wondering if you're still looking for this usecase and would be interested in evaluating it?

**Comment by jonathanq** [comment]
@alisaifee Yes - still interested in this usecase.  I see some commits about swallowing any errors from the backend when doing a limit.  Is that part of it?  I think that combined with using an in-memory store in the event of a backend failure would solve the problem completely.

**Comment by alisaifee** [comment]
@jonathanq: not exactly related, but that usecase did remind me of this outstanding issue.
the implementation for the in memory fallback can be found here at 5d58d1a92d1653691a879388c886b4ffc771d180, and specifically the test case [here](https://github.com/alisaifee/flask-limiter/commit/5d58d1a92d1653691a879388c886b4ffc771d180#diff-09f7c67f2044e4b0422a2fe397cb7e24R300) demonstrates the expected use case

**Comment by jonathanq** [comment]
I finally had a chance to try this out (very sorry for the delay!)

I used a remote Redis server and in-memory fall back enabled.  When I removed the connection to the remote Redis server, the request failed because of:

```
Traceback (most recent call last):
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/werkzeug/wsgi.py", line 648, in __call__
    return app(environ, start_response)
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1379, in handle_user_exception
    return handler(e)
  File "/Users/testuser/Documents/development/git/fillz-python/fillz/frontend/app.py", line 213, in exception_handler
    raise error
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1473, in full_dispatch_request
    rv = self.preprocess_request()
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/flask/app.py", line 1666, in preprocess_request
    rv = func()
  File "/Users/testuser/.virtualenvs/sampleapp/src/flask-limiter/flask_limiter/extension.py", line 345, in __check_request_limit
    six.reraise(*sys.exc_info())
  File "/Users/testuser/.virtualenvs/sampleapp/lib/python3.4/site-packages/six.py", line 659, in reraise
    raise value
  File "/Users/testuser/.virtualenvs/sampleapp/src/flask-limiter/flask_limiter/extension.py", line 300, in __check_request_limit
    if self.__should_check_backend() and self._storage.check():
AttributeError: 'RedisStorage' object has no attribute 'check'
```

Because the ```self._storage.check()``` fails - it never checks the in_memory check in this block of code in https://github.com/alisaifee/flask-limiter/blob/memory-fallback/flask_limiter/extension.py#L300:
```
all_limits = []
if self._storage_dead and self._fallback_limiter:
    if self.__should_check_backend() and self._storage.check():
        self._storage_dead = False
    else:
        all_limits = self._in_memory_fallback
```

Maybe this is just an issue with RedisStorage?

**Comment by jonathanq** [comment]
Forgot to add, this is how I have the app configured:

```
RATELIMIT_HEADERS_ENABLED = False
RATELIMIT_STORAGE_URL = "redis://localhost:6379"
RATELIMIT_IN_MEMORY_FALLBACK = "1/minute"
```

And the route had: ```@limiter.limit("5/minute")```

**Comment by alisaifee** [comment]
Hi @jonathanq. You'll have to update your version of the underlying `limits` dependency so that it contains [this](https://github.com/alisaifee/limits/commit/a3df0f7deebdfc855fe1cee1133ff5f50a3f1cd1) commit.

**Comment by jonathanq** [comment]
Pulled in that version and it worked perfectly.  

One suggestion I have is to add a message to the logger that it's fallen back to the in-memory.  And another message that logs when it has restored connection.

Right now I just get the warning message that a request exceeded the rate (and I can see it is using the in-memory rate when the Redis server is down).  But no other indication that it switched to in-memory fallback.  Other than that - the feature looks good.

**Comment by alisaifee** [comment]
Thanks for testing this out. I've added some logging for when the limiter switches to in-memory and switches back. 

I'm just ironing out a few minor details in the implementation and will hopefully have a release by EOD with this feature.

**Comment by alisaifee** [comment]
Released in `0.9`. Thanks a lot for the ideas and help!
